### PR TITLE
Script to Shell(Run)

### PR DESCRIPTION
**Pull request by alexg9010**
Hey People,

Those changes allow us now to cirumvent come of the internal changes that snakemake hiddenly does with our R scripts. Please test and let me know if it does not work for you.

### CODE REVIEW

File: scripts/multireport_functions.R [hunk 1]
```diff
@@ -1,3 +1,52 @@
...
+argsDF <- as.data.frame(do.call("rbind", parseArgs(args)))
+argsL <- as.list(as.character(argsDF$V2))
+names(argsL) <- argsDF$V1
+
+## get deeper list elements 
+if(!is.null(argsL$report.params)) {
+  argsL$report.params <- jsonlite::fromJSON(argsL$report.params)
+}
+
+# print(argsL)
```
Comment (rekado) [inline scripts/multireport_functions.R]: I think this should be removed.

File: scripts/multireport_functions.R [hunk 2]
```diff
@@ -1,3 +1,52 @@
+# Parse Command Line Arguments --------------------------------------------
+
+## Collect arguments
+args <- commandArgs(TRUE)
+
+print(args)
```
Comment (rekado) [inline scripts/multireport_functions.R]: This can be removed.

File: scripts/multireport_functions.R [hunk 3]
```diff
@@ -251,21 +300,29 @@ render2multireport <- function(final_output,
...
+                   references = argsL$references,
+                   sessioninfo = argsL$sessioninfo)
+
+# finalReportDir = "Final_Report/"
+# 
+# render2Markdown(reportFile = normalizePath(snakemake@input[["template"]]),
+#                 outFile = basename(snakemake@output[["report"]]),
+#                 outDir = normalizePath(dirname(snakemake@output[["report"]])),
+#                 finalReportDir = finalReportDir,
+#                 report.params = snakemake@params[nchar(names(snakemake@params)) > 0] )
```
Comment (rekado) [inline scripts/multireport_functions.R]: Please remove the commented stuff.  We can always access old code via git.

File: scripts/report_functions.R [hunk 4]
```diff
@@ -85,25 +132,31 @@ render2Markdown <- function(reportFile,
...
 sink(out, type = "message")
 
-cat(paste(  
-  Sys.time(),"\n\n",
-  "Rendering report:",basename(snakemake@output[["report"]]),"\n",
-  "from template:",basename(snakemake@input[["template"]]),"\n",
-  "into directory:",normalizePath(dirname(snakemake@output[["report"]])),"\n\n"
-  ))
+## debugging
+# save.image(file = "snakemakeObj.RData")
```
Comment (rekado) [inline scripts/report_functions.R]: Please remove this.

**Comment by rekado** [comment]
Good job!

I'm not really happy with the code duplication, though.  Could you describe what problems you had with using a procedure that takes rule state (wildcards, inputs, params, etc) and produces a command string?

**Comment by alexg9010** [comment]
Cool thanks!

Yes, I also do not like the repetition of this, 
but my problem was that if would use a procedure like this:
```
def makeCMD(input,output,params,wildcards,log):
  
        dumps = json.dumps(dict(params.items()),sort_keys=True,
                           separators=(",",":"), ensure_ascii=True)
                           
        cmd =   "{RSCRIPT} {DIR_scripts}/report_functions.R"
        cmd +=  " --reportFile={input.template}"
        cmd +=  " --outFile={output.report}"
        cmd +=  " --finalReportDir=" + os.path.join(DIR_final,wildcards.prefix)
        cmd +=  " --report.params={dumps:q}"
        cmd +=  " --logFile={log}" 

  return(cmd)
```
Then the string interpolation would not work and therewith the quote escaping (`{dumps:q}`) would fail.

**Comment by alexg9010** [comment]
So I was just now trying a new way:

This would be the procedure:
```
def makeCMD(input,output,params,knitDir,log):
    ## dumping twice to escape double quotes
    dumps = json.dumps(json.dumps(dict(params.items()),sort_keys=True,
                       separators=(",",":"), ensure_ascii=True))
                       
    cmd = [RSCRIPT,os.path.join(DIR_scripts,"report_functions.R"),
     "--reportFile="+input.template,
     "--outFile="+output.report,
     "--finalReportDir="+os.path.join(DIR_final,knitDir),
     "--report.params="+dumps,
     "--logFile="+str(log) ]
     
    cmd =  ' '.join(cmd)
    
    return(cmd)
```

which is called like this:
```
[...]
    run:
        cmd = makeCMD(input,output,params,wildcards.prefix,log)
        
        shell(cmd)
```

but this procedure leads to this error:
```
NameError in line 48 of /home/agosdsc/projects/pigx_bsseq/git/pigx_bsseq/rules/post_mapping.rules:
The name '\\"assembly\\"' is unknown in this context. Please make sure that you defined that variable. Also note that braces not used for variable access have to be escaped by repeating them, i.e. {{print $1}}
```

**Comment by alexg9010** [comment]
I guess this comes from the shell() command removing the escapes from the quotes and trying to interpolate the cmd string, which is actually not needed anymore. 

So the reason for keeping the  duplicated cmd pastings would be to have the option to see the executed command upon calling `./pigx-bsseq --snakeparams " -p" test/samplesheet.csv`.  

If we do not need this feature we could just use python subprocess.call() which takes care of escaping after we did interpolation manually.

**Comment by rekado** [comment]
alexg9010 <notifications@github.com> writes:

> Yes, I also do not like the repetition of this,
> but my problem was that if would use a procedure like this:
> ```
> def makeCMD(input,output,params,wildcards,log):
>
>         dumps = json.dumps(dict(params.items()),sort_keys=True,
>                            separators=(",",":"), ensure_ascii=True)
>
>         cmd =   "{RSCRIPT} {DIR_scripts}/report_functions.R"
>         cmd +=  " --reportFile={input.template}"
>         cmd +=  " --outFile={output.report}"
>         cmd +=  " --finalReportDir=" + os.path.join(DIR_final,wildcards.prefix)
>         cmd +=  " --report.params={dumps:q}"
>         cmd +=  " --logFile={log}"
>
>   return(cmd)
> ```
> Then the string interpolation would not work and therewith the quote escaping (`{dumps:q}`) would fail.

String interpolation happens in snakemake‘s `shell` procedure, so the
only real problem here is that we cannot generate a value for `dumps` in
the procedure.

`shell` takes more than one argument, though, so we could supply a value
for `dumps`.  And maybe that value can be computed by a procedure that
accepts `params`.

--
Ricardo

GPG: BCA6 89B6 3655 3801 C3C6  2150 197A 5888 235F ACAC
https://elephly.net

**Comment by rekado** [comment]
I'm using this procedure

```
def generateReport(input, output, params, log, reportSubDir):
    dumps = json.dumps(dict(params.items()),sort_keys=True,
                       separators=(",",":"), ensure_ascii=True)

    cmd =   "{RSCRIPT} {DIR_scripts}/report_functions.R"
    cmd +=  " --reportFile={input.template}"
    cmd +=  " --outFile={output.report}"
    cmd +=  " --finalReportDir=" + os.path.join(DIR_final,reportSubDir)
    cmd +=  " --report.params={dumps:q}"
    cmd +=  " --logFile={log}"
    shell(cmd, dumps)
```

The only downside is that it won't print the shell command before executing it (when "-p" is passed to snakemake).

**Pull request merged**